app/scripts/metamask-controller.js

```diff
@@ -1312,6 +1297,176 @@ export default class MetamaskController extends EventEmitter {
       }),
     });
 
+    this.txController.on(
```

Comment:
As this is lots of logic in an already big class, can we encapsulate all this inside functions in something like `app/scripts/lib/transaction-metrics.ts`?

It would also make it much easier to unit test.